agent/structs/structs.go

```diff
@@ -283,6 +288,12 @@ type QueryMeta struct {
 	// Having `discovery_max_stale` on the agent can affect whether
 	// the request was served by a leader.
 	ConsistencyLevel string
+
+	// NotModified is true when the Index of the query is the same value as the
+	// requested MinIndex. It indicates that the entity has not been modified.
+	// When NotModified is true, the response will not contain the result of
+	// the query.
```

Comment:
The caller could also infer that from the index in the response. Why do we need an extra field here?
I imagine that we have it to avoid ambiguity, but that could also be solved with a version check like: if server.version > 1.8.0 and req.index == resp.index, assume NotModified.
In addition to that, the client still has to explicitly request this behavior with `AllowNotModifiedResponse`.

Reply:
ahh, that's a good point. If we removed this field the client would know if it had set `AllowNotModifiedResponse` and as long as error is nil, and `Index == MinIndex` it could assume that an empty list indicates "Not Modified". The only time that wouldn't be true would be when the API is an older version without this support.

I think in general doing capability checks with application version is risky. It limits what you can change in the future. An explicit declaration of support for a feature is going to be a lot more reliable.

In this case, I guess the `NotModified` field is effectively the declaration that the feature is supported. I agree that it's only necessary for an edge case, but I'm tempted to keep it to be more explicit about what the response contains.

Reply:
Happy either way.

Reply:
@i0rek the flag is necessary because to me:
 * some endpoints might choose not to implement this optimization, and it allows client to disambiguate between empty collection in result OR empty result because no content
 * knowing the version or real server might be hard (think call forwarded to other DC/or to Leader)
 * some endpoints might choose to return results because we forgot something (think weird ACL change in the meantime, weird filter that returns nodes always changing...) and not letting the client infer this is important to me because empty collection or null might be a legit result

Reply:
@pierresouchay good points, I agree it is needed.